### ISSUE TITLE
If I paste RGB color, resulting oklch has too much precision and doesn’t fit into textbox

### ISSUE THREAD

**Issue by tonsky** [evilmartians/oklch-picker#158]
Steps to reproduce:

1. Paste `0067E2` into RGB

Expected:

- OkLCH has something like `oklch(54% 0.20 258)`

Actual:

- OkLCH has `oklch(54.12% 0.203760003850602 258.38835815691493)`

![Screenshot 2024-09-11 at 21 03 17](https://github.com/user-attachments/assets/b1605d7e-58fc-43d7-a78d-76c2f150093d)

**Comment by ai** [comment]
We do it for a reason. `#0067E2` is on the edge of sRGB gamut and with less precision the color will go to P3 gamut.

![Captura desde 2024-09-12 01-14-49](https://github.com/user-attachments/assets/d11debb9-7644-4485-878d-a9d2014335fa)

For instance, [`#0167e0`](https://oklch.com/#53.96,0.201,258.18,100) does not change gamut after the rounding the numbers, so we show short numbers for it.

So we check does the precision reduction change the gamut and if to change it, we do not round the numbers.

Here is the code https://github.com/evilmartians/oklch-picker/blob/36cd23902e9e52e2369337a8ba3a443441670e42/stores/current.ts#L189-L194

Any idea how we can improve the algorithm?

**Comment by tonsky** [comment]
Yes. Round the numbers. If it goes into P3 territory and display can’t show it, it will be rounded down anyway

**Comment by tonsky** [comment]
Is that bad? E.g. if I paste `#0067E2` into RGB, the color you show in preview is `#0167E2`. So there’s already an error happening. I think we all expect that color conversions come with a small amount of errors, at least because of precision

**Comment by ai** [comment]
Here we will have double shift. The second problem is that browser's P3 > sRGB gamut correction could be different leading to a different shift.

**Comment by ai** [comment]
Another problem is that it is confuse users. They put obviously sRGB color, but tool tell them that it is P3 color.

We added this feature because people asked us in another issue.

**Comment by SomeNameDee** [comment]
One idea could be, instead of rounding being all-or-nothing ("if the rounded value is a mismatch then don't round at all"), instead round incrementally.
As in, if the initial rounding fails to match, add an additional decimal place and try again until the rounding matches.
Just not sure how much slower this would be; if it's too dramatic a slowdown it might have to be gated behind a config flag or scrapped entirely.

**Comment by ai** [comment]
I like the idea. Can you send PR?

**Comment by ankiiisharma** [comment]
hey @ai , would like to work on this. Let me know if this issue is still open. Thanks.

**Comment by ai** [comment]
@ankiiisharma yes, we still need help here to implement @SomeNameDee algorithm

**Comment by ai** [comment]
I made a fix which make it a little better https://github.com/evilmartians/oklch-picker/commit/57ef89eae12749a1c34b7b0764215d5966bc24dc

It is still longer that field but at least is not soo long.
